### PR TITLE
Use new Kokkos occupancy feature for Cuda traversal

### PR DESCRIPTION
**Pull request by aprokop**
Couple things about the implementation.

First, we could start requiring Kokkos 3.3 after creating a 3.2 tag. I would be OK with it, but I would also be OK allowing older versions of Kokkos. Thus, at this point I protected it by the `KOKKOS_VERSION`, which first appeared in 3.1.00 which we require.

Second, I decided to only enable it for Cuda. I know that it's a no-op for others, but I don't want a surprise in the future once it's implemented for HIP. I also don't know what HIP parameters would need to be.

Last, I'm running some performance benchmarks with 25, 50 and 75 occupancy for Cuda on V100. I will post the results to figure out what is the best choice.

### CODE REVIEW

**Comment by aprokop** [comment]
To test this, we would need to switch one of the builds to Kokkos 3.3, which I think we should do anyways.

File: src/details/ArborX_DetailsTreeTraversal.hpp [hunk 1]
```diff
@@ -66,10 +66,20 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
               std::is_same<typename Node::Tag, NodeWithLeftChildAndRopeTag>{},
           "Unrecognized node tag");
 
-      Kokkos::parallel_for("ArborX::TreeTraversal::spatial",
-                           Kokkos::RangePolicy<ExecutionSpace>(
-                               space, 0, Access::size(predicates)),
-                           *this);
+#if defined(KOKKOS_ENABLE_CUDA) && KOKKOS_VERSION >= 30300
```
Comment (dalg24) [inline src/details/ArborX_DetailsTreeTraversal.hpp:69]: Please add a comment to capture what you discuss in the description of the PR, that is why you chose this even though you could in principle have expressed the preference for all execution spaces.

**Comment by aprokop** [comment]
I'm not sure what's going on, but the results are not pretty. In short, knn is OKish, with 25% being the fastest. However, radius is all over the place, and sometimes there is no difference. 

[master.json.txt](https://github.com/arborx/ArborX/files/5756333/master.json.txt)
[occupancy_25.json.txt](https://github.com/arborx/ArborX/files/5756334/occupancy_25.json.txt)
[occupancy_50.json.txt](https://github.com/arborx/ArborX/files/5756335/occupancy_50.json.txt)
[occupancy_75.json.txt](https://github.com/arborx/ArborX/files/5756336/occupancy_75.json.txt)

DBSCAN results seem to reproduce the 25% drop with 75% desired occupancy.

**Comment by aprokop** [comment]
Rebased on `master`.

**Comment by aprokop** [comment]
Plotted the results comparing `master`, `25%`, `50%`, 75%.

### Sorted queries
[occupancy_sorted.pdf](https://github.com/arborx/ArborX/files/5922258/occupancy_sorted.pdf)

Specifying one of these occupancies is always faster than master. The three behave very similar for the filled case, with slight edge to 25% for kNN. For the hollow case, though, 25% drastically outperforms the other two. I think finer grained exploration should be performed (say, 5%, 10%, 20%, 30%, 40%) to figure out the sweet spot.

### Unsorted queries
[occupancy_unsorted.pdf](https://github.com/arborx/ArborX/files/5922259/occupancy_unsorted.pdf)

Mixed bag. For the filled case, the occupancies mostly help, with 25% being the winner (though master is faster for some specific sizes, like 10K for spatial). For the hollow case, there is parity between master and any occupancy for kNN, but master dramatically wins for the hollow spatial for all problems sizes, almost by a factor of 2.

### What's next

I think the results should be studied further. There seem to be many situations where not sorting queries is actually significantly faster than sorting them. But if, for a moment, we say we don't care about unsorted queries, the path forward is pretty straightforward: run a few more occupancy values at the lower end [5-40] and see how sensitive the code to that is.

**Comment by aprokop** [comment]
### Sorted vs unsorted

Darn. Did not expect this.

[sorted_vs_unsorted_master.pdf](https://github.com/arborx/ArborX/files/5922372/sorted_vs_unsorted_master.pdf)

**Comment by aprokop** [comment]
I'm not sure if we'll ever get back to it. I see no value in keeping it open, it's been year+ since any real work have been tried. Will keep the branch around, but closing.

**Pull request closed without merging**